File: backend/app/services/timesheet_service.py
"""Сервис трудозатрат: ввод, недельная сетка, отправка и утверждение.

Ключевые инварианты (CLAUDE.md §4):
- часы шаг 0.25, не более 24/день суммарно, комментарий обязателен;
- утверждает руководитель проекта; после утверждения запись блокируется;
- корректировка утверждённой записи — сторно (reversal), не редактирование;
- себестоимость берётся из ставки, действующей на дату записи (снимок при утв.).
"""

from __future__ import annotations

import uuid
from datetime import UTC, date, datetime, timedelta
from decimal import Decimal

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import (
    AppError,
    ConflictError,
    ErrorCode,
    ForbiddenError,
    NotFoundError,
    ValidationError,
)
from app.models.enums import AuditAction, Role, TimeEntryStatus
from app.models.project import Project
from app.models.timesheet import TimeEntry, TimesheetWeek
from app.models.user import User
from app.repositories.timesheet_repo import TimesheetRepository
from app.repositories.user_repo import UserRepository
from app.schemas.timesheet import TimeEntryCreate, TimeEntryUpdate
from app.services import timesheet_rules
from app.services.audit_service import AuditService

_EDITABLE = {TimeEntryStatus.DRAFT, TimeEntryStatus.REJECTED}


def monday_of(d: date) -> date:
    return d - timedelta(days=d.weekday())


class TimesheetService:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session
        self.repo = TimesheetRepository(session)
        self.users = UserRepository(session)
        self.audit = AuditService(session)

    # ---------- Недельная сетка ----------
    async def get_week(self, user: User, week_start: date) -> dict:
        start = monday_of(week_start)
        end = start + timedelta(days=6)
        entries = list(await self.repo.list_for_week(user.id, start, end))
        week = await self.repo.get_week(user.id, start)
        daily: dict[str, Decimal] = {}
        total = Decimal("0")
        for e in entries:
            total += e.hours
            key = e.work_date.isoformat()
            daily[key] = daily.get(key, Decimal("0")) + e.hours
        return {
            "week_start": start,
            "week_end": end,
            "status": week.status if week else TimeEntryStatus.DRAFT,
            "entries": entries,
            "total_hours": total,
            "daily_totals": daily,
        }

    # ---------- Ввод ----------
    async def _ensure_can_log(self, user: User, project_id: uuid.UUID) -> None:
        if not await self.repo.is_project_member(project_id, user.id):
            raise ForbiddenError("Вы не участник проекта — списание часов недоступно")

    async def create_entry(self, user: User, data: TimeEntryCreate) -> TimeEntry:
        await self._ensure_can_log(user, data.project_id)
        try:
            timesheet_rules.validate_hours(data.hours)
        except ValueError as exc:
            raise ValidationError(str(exc), code=ErrorCode.TIMESHEET_INVALID_HOURS) from exc
        existing = await self.repo.daily_total(user.id, data.work_date)
        try:
            timesheet_rules.validate_daily_total(existing, data.hours)
        except ValueError as exc:
            raise ValidationError(str(exc), code=ErrorCode.TIMESHEET_DAILY_LIMIT) from exc

        entry = await self.repo.create(
            user_id=user.id,
            project_id=data.project_id,
            task_id=data.task_id,
            work_date=data.work_date,
            hours=data.hours,
            comment=data.comment,
            status=TimeEntryStatus.DRAFT,
        )
        return entry

    async def _own_editable(self, user: User, entry_id: uuid.UUID) -> TimeEntry:
        entry = await self.repo.get(entry_id)
        if entry is None:
            raise NotFoundError("Запись не найдена")
        if entry.user_id != user.id:
            raise ForbiddenError("Можно редактировать только свои записи")
        if entry.status not in _EDITABLE:
            raise AppError(
                "Запись отправлена или утверждена — редактирование запрещено "
                "(используйте сторно для утверждённых)",
                code=ErrorCode.TIMESHEET_LOCKED,
                status_code=409,
            )
        return entry

    async def update_entry(
        self, user: User, entry_id: uuid.UUID, data: TimeEntryUpdate
    ) -> TimeEntry:
        entry = await self._own_editable(user, entry_id)
        new_hours = data.hours if data.hours is not None else entry.hours
        if data.hours is not None:
            try:
                timesheet_rules.validate_hours(new_hours)
            except ValueError as exc:
                raise ValidationError(
                    str(exc), code=ErrorCode.TIMESHEET_INVALID_HOURS
                ) from exc
            existing = await self.repo.daily_total(
                user.id, entry.work_date, exclude_id=entry.id
            )
            try:
                timesheet_rules.validate_daily_total(existing, new_hours)
            except ValueError as exc:
                raise ValidationError(
                    str(exc), code=ErrorCode.TIMESHEET_DAILY_LIMIT
                ) from exc
            entry.hours = new_hours
        if data.task_id is not None:
            entry.task_id = data.task_id
        if data.comment is not None:
            entry.comment = data.comment
        # Повторная отправка после отклонения возвращает в черновик.
        if entry.status == TimeEntryStatus.REJECTED:
            entry.status = TimeEntryStatus.DRAFT
            entry.reject_reason = None
        await self.session.flush()
        return entry

    async def delete_entry(self, user: User, entry_id: uuid.UUID) -> None:
        entry = await self._own_editable(user, entry_id)
        await self.repo.soft_delete(entry)

    # ---------- Отправка ----------
    async def submit_week(self, user: User, week_start: date) -> int:
        start = monday_of(week_start)
        end = start + timedelta(days=6)
        drafts = list(await self.repo.get_entries_for_week_submit(user.id, start, end))
        if not drafts:
            raise ValidationError("Нет черновиков для отправки на этой неделе")
        for e in drafts:
            e.status = TimeEntryStatus.SUBMITTED
        week = await self.repo.get_week(user.id, start)
        if week is None:
            week = TimesheetWeek(
                user_id=user.id, week_start=start, status=TimeEntryStatus.SUBMITTED
            )
            self.session.add(week)
        else:
            week.status = TimeEntryStatus.SUBMITTED
        await self.session.flush()
        return len(drafts)

    async def copy_week(
        self, user: User, source_week_start: date, target_week_start: date
    ) -> int:
        src = monday_of(source_week_start)
        tgt = monday_of(target_week_start)
        if src == tgt:
            raise ValidationError("Неделя-источник совпадает с целевой")
        src_entries = list(await self.repo.list_for_week(user.id, src, src + timedelta(days=6)))
        if not src_entries:
            raise ValidationError("В неделе-источнике нет записей для копирования")
        offset = (tgt - src).days
        created = 0
        for e in src_entries:
            new_date = e.work_date + timedelta(days=offset)
            existing = await self.repo.daily_total(user.id, new_date)
            if existing + e.hours > timesheet_rules.MAX_DAILY_HOURS:
                continue  # не превышаем суточный лимит — пропускаем
            await self.repo.create(
                user_id=user.id,
                project_id=e.project_id,
                task_id=e.task_id,
                work_date=new_date,
                hours=e.hours,
                comment=e.comment,
                status=TimeEntryStatus.DRAFT,
            )
            created += 1
        return created

    # ---------- Утверждение ----------
    @staticmethod
    def _is_privileged(approver: User) -> bool:
        return Role.ADMIN in set(approver.roles)

    async def _assert_can_approve(self, approver: User, entry: TimeEntry) -> None:
        if self._is_privileged(approver):
            return
        project = await self.session.get(Project, entry.project_id)
        if project is None:
            raise NotFoundError("Проект записи не найден")
        if project.manager_id != approver.id and project.curator_id != approver.id:
            raise AppError(
                "Утверждать может только руководитель/куратор проекта",
                code=ErrorCode.NOT_APPROVER,
                status_code=403,
            )

    async def pending_for(self, approver: User) -> list[TimeEntry]:
        return list(
            await self.repo.pending_for_approver(
                approver.id, is_privileged=self._is_privileged(approver)
            )
        )

    async def approve(self, approver: User, entry_ids: list[uuid.UUID]) -> int:
        approved = 0
        touched_projects: set[uuid.UUID] = set()
        for eid in entry_ids:
            entry = await self.repo.get(eid)
            if entry is None or entry.status != TimeEntryStatus.SUBMITTED:
                continue
            await self._assert_can_approve(approver, entry)
            rate_row = await self.users.effective_cost_rate(entry.user_id, entry.work_date)
            entry.status = TimeEntryStatus.APPROVED
            entry.approved_by = approver.id
            entry.approved_at = datetime.now(UTC)
            entry.cost_rate_snapshot = rate_row.cost_rate if rate_row else Decimal("0")
            touched_projects.add(entry.project_id)
            approved += 1
        if approved:
            await self.audit.record(
                entity="TimeEntry",
                entity_id=None,
                action=AuditAction.APPROVE,
                actor_id=approver.id,
                diff={"count": approved, "ids": [str(i) for i in entry_ids]},
            )
            await self.session.flush()
            self._schedule_margin_recalc(touched_projects)
        return approved

    async def reject(
        self, approver: User, entry_ids: list[uuid.UUID], reason: str | None
    ) -> int:
        if not reason or not reason.strip():
            raise ValidationError("Отклонение требует указания причины")
        from app.services.notification_service import NotificationService

        notifier = NotificationService(self.session)
        rejected = 0
        for eid in entry_ids:
            entry = await self.repo.get(eid)
            if entry is None or entry.status != TimeEntryStatus.SUBMITTED:
                continue
            await self._assert_can_approve(approver, entry)
            entry.status = TimeEntryStatus.REJECTED
            entry.reject_reason = reason
            await notifier.notify(
                entry.user_id,
                kind="timesheet_rejected",
                title="Трудозатраты отклонены",
                body=f"Запись от {entry.work_date.isoformat()} отклонена: {reason}",
                link="/timesheets",
            )
            rejected += 1
        if rejected:
            await self.audit.record(
                entity="TimeEntry",
                entity_id=None,
                action=AuditAction.REJECT,
                actor_id=approver.id,
                diff={"count": rejected, "reason": reason},
            )
            await self.session.flush()
        return rejected

    async def reverse(self, approver: User, entry_id: uuid.UUID) -> TimeEntry:
        """Сторно утверждённой записи: создаёт зеркальную запись с отрицательными
        часами и снимком стоимости, чтобы откатить себестоимость без правки оригинала."""
        entry = await self.repo.get(entry_id)
        if entry is None:
            raise NotFoundError("Запись не найдена")
        if entry.status != TimeEntryStatus.APPROVED:
            raise ConflictError("Сторнировать можно только утверждённую запись")
        await self._assert_can_approve(approver, entry)
        reversal = await self.repo.create(
            user_id=entry.user_id,
            project_id=entry.project_id,
            task_id=entry.task_id,
            work_date=entry.work_date,
            hours=-entry.hours,
            comment=f"Сторно записи от {entry.work_date.isoformat()}: {entry.comment}",
            status=TimeEntryStatus.APPROVED,
            approved_by=approver.id,
            approved_at=datetime.now(UTC),
            cost_rate_snapshot=entry.cost_rate_snapshot,
            reversal_of=entry.id,
        )
        await self.audit.record(
            entity="TimeEntry",
            entity_id=entry.id,
            action=AuditAction.UPDATE,
            actor_id=approver.id,
            diff={"reversal": str(reversal.id)},
        )
        self._schedule_margin_recalc({entry.project_id})
        return reversal

    @staticmethod
    def _schedule_margin_recalc(project_ids: set[uuid.UUID]) -> None:
        """Поставить пересчёт маржи в Celery (best-effort: брокер может быть недоступен)."""
        from app.core.config import settings

        if not settings.tasks_enabled:
            return
        try:
            from app.tasks.jobs import recalc_project_margin

            for pid in project_ids:
                recalc_project_margin.delay(str(pid))
        except Exception:
            pass

File: backend-dotnet/src/Veha.Api/Services/TimesheetService.cs
using Microsoft.EntityFrameworkCore;
using Veha.Api.Dtos;
using Veha.Domain.Common;
using Veha.Domain.Entities;
using Veha.Domain.Enums;
using Veha.Domain.Rules;
using Veha.Infrastructure;

namespace Veha.Api.Services;

/// <summary>Трудозатраты: недельная сетка, ввод, отправка, утверждение, сторно.
/// Порт services/timesheet_service.py. Себестоимость фиксируется снимком ставки на
/// дату записи при утверждении; корректировка утверждённого — сторно, не правка.</summary>
public class TimesheetService(VehaDbContext db, AuditService audit, NotificationService notifications)
{
    private static readonly TimeEntryStatus[] Editable = [TimeEntryStatus.Draft, TimeEntryStatus.Rejected];

    private static DateOnly MondayOf(DateOnly d)
        => d.AddDays(-(((int)d.DayOfWeek + 6) % 7));

    // ---------- Недельная сетка ----------
    public async Task<WeekResponseDto> GetWeekAsync(User user, DateOnly weekStart, CancellationToken ct)
    {
        var start = MondayOf(weekStart);
        var end = start.AddDays(6);
        var entries = await db.TimeEntries
            .Where(e => e.UserId == user.Id && e.WorkDate >= start && e.WorkDate <= end)
            .OrderBy(e => e.WorkDate)
            .ToListAsync(ct);
        var week = await db.TimesheetWeeks.FirstOrDefaultAsync(w => w.UserId == user.Id && w.WeekStart == start, ct);

        decimal total = 0;
        var daily = new Dictionary<string, decimal>();
        foreach (var e in entries)
        {
            total += e.Hours;
            var key = e.WorkDate.ToString("yyyy-MM-dd");
            daily[key] = daily.GetValueOrDefault(key) + e.Hours;
        }

        return new WeekResponseDto(start, end, week?.Status ?? TimeEntryStatus.Draft,
            entries.Select(ToOut).ToList(), total, daily);
    }

    // ---------- Ввод ----------
    private async Task EnsureCanLogAsync(User user, Guid projectId, CancellationToken ct)
    {
        var member = await db.ProjectMembers.AnyAsync(m => m.ProjectId == projectId && m.UserId == user.Id, ct);
        if (member) return;
        var lead = await db.Projects.AnyAsync(
            p => p.Id == projectId && (p.ManagerId == user.Id || p.CuratorId == user.Id), ct);
        if (!lead)
            throw new ForbiddenException("Вы не участник проекта — списание часов недоступно");
    }

    public async Task<TimeEntryOutDto> CreateEntryAsync(User user, TimeEntryCreateDto dto, CancellationToken ct)
    {
        await EnsureCanLogAsync(user, dto.ProjectId, ct);
        try { TimesheetRules.ValidateHours(dto.Hours); }
        catch (DomainValidationException ex) { throw new TimesheetInvalidHoursException(ex.Message); }

        var existing = await DailyTotalAsync(user.Id, dto.WorkDate, null, ct);
        try { TimesheetRules.ValidateDailyTotal(existing, dto.Hours); }
        catch (DomainValidationException ex) { throw new TimesheetDailyLimitException(ex.Message); }

        var entry = new TimeEntry
        {
            UserId = user.Id, ProjectId = dto.ProjectId, TaskId = dto.TaskId,
            WorkDate = dto.WorkDate, Hours = dto.Hours, Comment = dto.Comment,
            Status = TimeEntryStatus.Draft,
        };
        db.TimeEntries.Add(entry);
        await db.SaveChangesAsync(ct);
        return ToOut(entry);
    }

    private async Task<TimeEntry> OwnEditableAsync(User user, Guid entryId, CancellationToken ct)
    {
        var entry = await db.TimeEntries.FirstOrDefaultAsync(e => e.Id == entryId, ct)
                    ?? throw new NotFoundException("Запись не найдена");
        if (entry.UserId != user.Id)
            throw new ForbiddenException("Можно редактировать только свои записи");
        if (!Editable.Contains(entry.Status))
            throw new TimesheetLockedException(
                "Запись отправлена или утверждена — редактирование запрещено (используйте сторно для утверждённых)");
        return entry;
    }

    public async Task<TimeEntryOutDto> UpdateEntryAsync(User user, Guid entryId, TimeEntryUpdateDto dto, CancellationToken ct)
    {
        var entry = await OwnEditableAsync(user, entryId, ct);
        if (dto.Hours is not null)
        {
            var newHours = dto.Hours.Value;
            try { TimesheetRules.ValidateHours(newHours); }
            catch (DomainValidationException ex) { throw new TimesheetInvalidHoursException(ex.Message); }
            var existing = await DailyTotalAsync(user.Id, entry.WorkDate, entry.Id, ct);
            try { TimesheetRules.ValidateDailyTotal(existing, newHours); }
            catch (DomainValidationException ex) { throw new TimesheetDailyLimitException(ex.Message); }
            entry.Hours = newHours;
        }
        if (dto.TaskId is not null) entry.TaskId = dto.TaskId;
        if (dto.Comment is not null) entry.Comment = dto.Comment;
        // Повторная правка после отклонения возвращает в черновик.
        if (entry.Status == TimeEntryStatus.Rejected)
        {
            entry.Status = TimeEntryStatus.Draft;
            entry.RejectReason = null;
        }
        await db.SaveChangesAsync(ct);
        return ToOut(entry);
    }

    public async Task DeleteEntryAsync(User user, Guid entryId, CancellationToken ct)
    {
        var entry = await OwnEditableAsync(user, entryId, ct);
        entry.DeletedAt = DateTimeOffset.UtcNow;
        await db.SaveChangesAsync(ct);
    }

    // ---------- Отправка ----------
    public async Task<int> SubmitWeekAsync(User user, DateOnly weekStart, CancellationToken ct)
    {
        var start = MondayOf(weekStart);
        var end = start.AddDays(6);
        var drafts = await db.TimeEntries
            .Where(e => e.UserId == user.Id && e.WorkDate >= start && e.WorkDate <= end
                        && e.Status == TimeEntryStatus.Draft)
            .ToListAsync(ct);
        if (drafts.Count == 0)
            throw new DomainValidationException("Нет черновиков для отправки на этой неделе");

        foreach (var e in drafts) e.Status = TimeEntryStatus.Submitted;

        var week = await db.TimesheetWeeks.FirstOrDefaultAsync(w => w.UserId == user.Id && w.WeekStart == start, ct);
        if (week is null)
            db.TimesheetWeeks.Add(new TimesheetWeek { UserId = user.Id, WeekStart = start, Status = TimeEntryStatus.Submitted });
        else
            week.Status = TimeEntryStatus.Submitted;

        await db.SaveChangesAsync(ct);
        return drafts.Count;
    }

    public async Task<int> CopyWeekAsync(User user, DateOnly sourceWeekStart, DateOnly targetWeekStart, CancellationToken ct)
    {
        var src = MondayOf(sourceWeekStart);
        var tgt = MondayOf(targetWeekStart);
        if (src == tgt)
            throw new DomainValidationException("Неделя-источник совпадает с целевой");

        var srcEntries = await db.TimeEntries
            .Where(e => e.UserId == user.Id && e.WorkDate >= src && e.WorkDate <= src.AddDays(6))
            .ToListAsync(ct);
        if (srcEntries.Count == 0)
            throw new DomainValidationException("В неделе-источнике нет записей для копирования");

        var offset = tgt.DayNumber - src.DayNumber;
        var created = 0;
        foreach (var e in srcEntries)
        {
            var newDate = e.WorkDate.AddDays(offset);
            var existing = await DailyTotalAsync(user.Id, newDate, null, ct);
            if (existing + e.Hours > TimesheetRules.MaxDailyHours)
                continue;  // не превышаем суточный лимит — пропускаем
            db.TimeEntries.Add(new TimeEntry
            {
                UserId = user.Id, ProjectId = e.ProjectId, TaskId = e.TaskId,
                WorkDate = newDate, Hours = e.Hours, Comment = e.Comment, Status = TimeEntryStatus.Draft,
            });
            await db.SaveChangesAsync(ct);  // как per-create flush в Python: следующий daily_total видит копии
            created++;
        }
        return created;
    }

    // ---------- Утверждение ----------
    public async Task<List<PendingEntryDto>> PendingForAsync(User approver, CancellationToken ct)
    {
        var privileged = approver.Roles.Contains("admin");
        var q = db.TimeEntries.Where(e => e.Status == TimeEntryStatus.Submitted);
        if (!privileged)
            q = q.Where(e => db.Projects.Any(
                p => p.Id == e.ProjectId && (p.ManagerId == approver.Id || p.CuratorId == approver.Id)));
        var entries = await q.OrderBy(e => e.WorkDate).ToListAsync(ct);

        var userIds = entries.Select(e => e.UserId).Distinct().ToList();
        var projIds = entries.Select(e => e.ProjectId).Distinct().ToList();
        var names = await db.Users.Where(u => userIds.Contains(u.Id)).ToDictionaryAsync(u => u.Id, u => u.FullName, ct);
        var codes = await db.Projects.Where(p => projIds.Contains(p.Id)).ToDictionaryAsync(p => p.Id, p => p.Code, ct);

        return entries.Select(e => new PendingEntryDto(
            e.Id, e.UserId, e.ProjectId, e.TaskId, e.WorkDate, e.Hours, e.Comment, e.Status,
            e.ApprovedBy, e.ApprovedAt, e.RejectReason, e.ReversalOf,
            names.GetValueOrDefault(e.UserId, "—"), codes.GetValueOrDefault(e.ProjectId, "—"))).ToList();
    }

    private async Task AssertCanApproveAsync(User approver, TimeEntry entry, CancellationToken ct)
    {
        if (approver.Roles.Contains("admin")) return;
        var project = await db.Projects.FirstOrDefaultAsync(p => p.Id == entry.ProjectId, ct)
                      ?? throw new NotFoundException("Проект записи не найден");
        if (project.ManagerId != approver.Id && project.CuratorId != approver.Id)
            throw new NotApproverException("Утверждать может только руководитель/куратор проекта");
    }

    public async Task<int> ApproveAsync(User approver, List<Guid> entryIds, CancellationToken ct)
    {
        var approved = 0;
        foreach (var eid in entryIds)
        {
            var entry = await db.TimeEntries.FirstOrDefaultAsync(e => e.Id == eid, ct);
            if (entry is null || entry.Status != TimeEntryStatus.Submitted) continue;
            await AssertCanApproveAsync(approver, entry, ct);
            var rate = await EffectiveRateAsync(entry.UserId, entry.WorkDate, ct);
            entry.Status = TimeEntryStatus.Approved;
            entry.ApprovedBy = approver.Id;
            entry.ApprovedAt = DateTimeOffset.UtcNow;
            entry.CostRateSnapshot = rate ?? 0m;
            approved++;
        }
        if (approved > 0)
            await audit.RecordAsync("TimeEntry", null, AuditAction.Approve, approver.Id,
                new { count = approved, ids = entryIds.Select(i => i.ToString()).ToArray() }, ct);
        return approved;
    }

    public async Task<int> RejectAsync(User approver, List<Guid> entryIds, string? reason, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(reason))
            throw new DomainValidationException("Отклонение требует указания причины");

        var rejected = 0;
        foreach (var eid in entryIds)
        {
            var entry = await db.TimeEntries.FirstOrDefaultAsync(e => e.Id == eid, ct);
            if (entry is null || entry.Status != TimeEntryStatus.Submitted) continue;
            await AssertCanApproveAsync(approver, entry, ct);
            entry.Status = TimeEntryStatus.Rejected;
            entry.RejectReason = reason;
            await notifications.NotifyAsync(entry.UserId, "timesheet_rejected", "Трудозатраты отклонены",
                $"Запись от {entry.WorkDate:yyyy-MM-dd} отклонена: {reason}", "/timesheets", ct);
            rejected++;
        }
        if (rejected > 0)
            await audit.RecordAsync("TimeEntry", null, AuditAction.Reject, approver.Id,
                new { count = rejected, reason }, ct);
        return rejected;
    }

    public async Task<TimeEntryOutDto> ReverseAsync(User approver, Guid entryId, CancellationToken ct)
    {
        var entry = await db.TimeEntries.FirstOrDefaultAsync(e => e.Id == entryId, ct)
                    ?? throw new NotFoundException("Запись не найдена");
        if (entry.Status != TimeEntryStatus.Approved)
            throw new ConflictException("Сторнировать можно только утверждённую запись");
        await AssertCanApproveAsync(approver, entry, ct);

        var reversal = new TimeEntry
        {
            UserId = entry.UserId, ProjectId = entry.ProjectId, TaskId = entry.TaskId, WorkDate = entry.WorkDate,
            Hours = -entry.Hours,
            Comment = $"Сторно записи от {entry.WorkDate:yyyy-MM-dd}: {entry.Comment}",
            Status = TimeEntryStatus.Approved, ApprovedBy = approver.Id, ApprovedAt = DateTimeOffset.UtcNow,
            CostRateSnapshot = entry.CostRateSnapshot, ReversalOf = entry.Id,
        };
        db.TimeEntries.Add(reversal);
        await audit.RecordAsync("TimeEntry", entry.Id, AuditAction.Update, approver.Id,
            new { reversal = reversal.Id.ToString() }, ct);
        return ToOut(reversal);
    }

    // ---------- Вспомогательное ----------
    private async Task<decimal> DailyTotalAsync(Guid userId, DateOnly workDate, Guid? excludeId, CancellationToken ct)
    {
        var q = db.TimeEntries.Where(e => e.UserId == userId && e.WorkDate == workDate);
        if (excludeId is not null) q = q.Where(e => e.Id != excludeId);
        return await q.SumAsync(e => (decimal?)e.Hours, ct) ?? 0m;
    }

    private async Task<decimal?> EffectiveRateAsync(Guid userId, DateOnly onDate, CancellationToken ct)
    {
        var rate = await db.UserCostRates
            .Where(r => r.UserId == userId && r.ValidFrom <= onDate && (r.ValidTo == null || r.ValidTo >= onDate))
            .OrderByDescending(r => r.ValidFrom)
            .FirstOrDefaultAsync(ct);
        return rate?.CostRate;
    }

    private static TimeEntryOutDto ToOut(TimeEntry e) => new(
        e.Id, e.UserId, e.ProjectId, e.TaskId, e.WorkDate, e.Hours, e.Comment, e.Status,
        e.ApprovedBy, e.ApprovedAt, e.RejectReason, e.ReversalOf);
}
